internal/transport/grpchttp2/http2bridge_test.go

```diff
@@ -0,0 +1,463 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpchttp2
+
+import (
+	"testing"
+
+	"golang.org/x/net/http2/hpack"
+)
+
+type fakeConn struct {
+	WBuf []byte
+	RBuf []byte
+}
+
+func (c *fakeConn) Write(p []byte) (int, error) {
+	c.WBuf = append(c.WBuf, p...)
+	return len(p), nil
+}
+
+func (c *fakeConn) Read(p []byte) (int, error) {
+	n := copy(p, c.RBuf)
+	c.RBuf = c.RBuf[n:]
+	return n, nil
+}
+
+func appendUint32(b []byte, x uint32) []byte {
+	return append(b, byte(x>>24), byte(x>>16), byte(x>>8), byte(x))
+}
+
+func readUint32(b []byte) uint32 {
+	return uint32(b[0])<<24 | uint32(b[1])<<16 | uint32(b[2])<<8 | uint32(b[3])
+}
+
+func readUint24(b []byte) int {
+	return int(b[0])<<16 | int(b[1])<<8 | int(b[2])
+}
+
+func readUint16(b []byte) uint16 {
+	return uint16(b[0])<<8 | uint16(b[1])
+}
+
+func checkWrittenHeader(t *testing.T, name string, hdr []byte, size int, ft FrameType, f Flag, streamID uint32) {
+	if gotSize := readUint24(hdr[0:3]); gotSize != size {
+		t.Errorf("%s(): Size: got %d, want %d", name, gotSize, size)
+	}
+	if gotType := FrameType(hdr[3]); gotType != ft {
+		t.Errorf("%s(): Type: got %#x, want %#x", name, gotType, ft)
+	}
+	if gotFlag := Flag(hdr[4]); gotFlag != f {
+		t.Errorf("%s(): Flags: got %#x, want %#x", name, gotFlag, f)
+	}
+	if gotSID := readUint32(hdr[5:9]); gotSID != streamID {
+		t.Errorf("%s(): StreamID: got %d, want %d", name, gotSID, streamID)
+	}
+}
+
+func checkReadHeader(t *testing.T, hdr *FrameHeader, size int, ft FrameType, f Flag, streamID uint32) {
+	if hdr.Size != uint32(size) {
+		t.Errorf("ReadFrame(): Size: got %d, want %d", hdr.Size, size)
+	}
+	if hdr.Type != ft {
+		t.Errorf("ReadFrame(): Type: got %#x, want %#x", hdr.Type, ft)
+	}
+	if hdr.Flags != f {
+		t.Errorf("ReadFrame(): Flags: got %#x, want %#x", hdr.Flags, f)
+	}
+	if hdr.StreamID != streamID {
+		t.Errorf("ReadFrame(): StreamID: got %d, want %d", hdr.StreamID, streamID)
+	}
+
+}
+
+func (s) TestHTTP2Bridge_ReadFrame_Data(t *testing.T) {
+	c := &fakeConn{}
+	recvData := "test data"
+	c.RBuf = append(c.RBuf, 0, 0, byte(len(recvData)), byte(FrameTypeData), byte(FlagDataEndStream))
+	c.RBuf = appendUint32(c.RBuf, 1)
+	c.RBuf = append(c.RBuf, []byte(recvData)...)
```

Comment:
Most of the tests have this boiler plate. Maybe make a helper function for this?

Reply:
@easwars  and I have discussed this before. We usually have setUp Test helpers, which helps remove loc but abstracts some of the setup stuff away.  I dont have a strong opinion on what the right tradeoff here is.

Reply:
That was my thought process as well, if I write a function that writes the bytes as a Frame wouldn't that be a framer itself? I wasn't sure if that kind of test was ok. I don't have a strong opinion either, if you think writing that sort of function still makes the test valid, I am up for it.